source/simulator/solver.cc

```diff
@@ -253,25 +253,84 @@ namespace aspect
       S_block_tolerance(S_block_tolerance)
     {}
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerA, class SInvOperator>
     unsigned int
-    BlockSchurPreconditioner<PreconditionerA, PreconditionerMp>::
+    BlockSchurPreconditioner<PreconditionerA, SInvOperator>::
     n_iterations_A() const
     {
       return n_iterations_A_;
     }
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerA, class SInvOperator>
     unsigned int
-    BlockSchurPreconditioner<PreconditionerA, PreconditionerMp>::
+    BlockSchurPreconditioner<PreconditionerA, SInvOperator>::
     n_iterations_S() const
     {
-      return n_iterations_S_;
+      return S_inv_operator.n_iterations;
     }
+    
+    template <class PreconditionerMp>
+    class InverseWeightedMassMatrix
+    {
+      public:
+        const TrilinosWrappers::SparseMatrix &mp_matrix;
+        PreconditionerMp &mp_preconditioner;
+        double S_block_tolerance;
```

Comment:
If all of these parameters are set in the constructor anyway they can probably be private as well?